### PR TITLE
Add VM instruction logging option

### PR DESCRIPTION
**Pull request by ixje**
**What current issue(s) does this address, or what feature is it adding?**
A continuation of #294 , specifically point 3

> add instruction logging
rationale; having a historical log of all calls executed gave me nearly all the information needed to exactly pin point which code line of my SC the VM was executing (by looking at function calls, DROP/RET statements etc).

**How did you solve this problem?**
Add some file writing statements to ApplicationEngine/ExecutionEngine and StateReader.

**How did you make sure your solution works?**
write tests + `make test`

**Are there any special changes in the code that we should be aware of?**
no

**Please check the following, if applicable:**

- [X] Did you add any tests?
- [X] Did you run `make lint`?
- [X] Did you run `make test`?
- [X] Are you making a PR to a feature branch or development rather than master?
- [X] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

### CODE REVIEW

**Comment by ixje** [comment]
Note: I specifically choose to implement the file writing in the `Execute()` function over the `StepInto()` function because the latter would mean you'd get hundreds of file open/closes per script execution (literally for every instruction).

File: neo/SmartContract/StateReader.py [hunk 1]
```diff
@@ -284,6 +284,9 @@ def Runtime_Notify(self, engine):
             payload
         )
 
+        message = payload[0] if len(payload) > 0 else payload
+        engine.write_log(str(message))
+
```
Comment (ixje) [inline neo/SmartContract/StateReader.py]: FYI; This and the one below is added to make sure that we don't just see a 
```
14 SYSCALL
Neo.Runtime.Notify
```
but also see the string/argument it's printing in the log
```
14 SYSCALL
Neo.Runtime.Notify
b'Start main'
```

**Comment by ixje** [comment]
The concern raised was speed, which is indeed important to keep in mind. Of course there's some impact to code additions, therefore I did a basic optimisation on the calls that get called _every_ instruction. This yielded a 9% improvement to offset the <2% loss by the new functionality.

**Comment by localhuman** [comment]
Looks good! Tests seem to be wanting `timeit` from the `neo.Utils` module

**Comment by localhuman** [comment]
This will be very helpful for many people until we get some actual debugging in place!

**Pull request merged**